system-configuration/src/network_configuration.rs

```diff
@@ -0,0 +1,414 @@
+    let path = CFString::from_static_string("State:/Network/Global/IPv4");
+    if let Some(value) = store.get(path.clone()) {
+        if let Some(dict) = value.downcast_into::<CFDictionary>() {
+            if let Some(val) = dict.find2(&CFString::new(key)) {
+                let value = unsafe { CFType::wrap_under_get_rule(val) };
+                if let Some(value) = value.downcast::<CFString>() {
+                    return Some(value.to_string());
+                }
+            }
+        }
+    }
+
+    return None;
+}
+
+
+/// Returns default route on primary network service.
+pub fn global_router(store: &SCDynamicStore) -> Option<IpAddr> {
+    if let Some(router_str) = global_query(store, "Router") {
+        if let Ok(router_ip) = router_str.parse::<IpAddr>() {
+            return Some(router_ip);
+        }
+    }
+
+    return None;
+}
+
+
+declare_TCFType!{
+    /// Network service object.
+    SCNetworkService, SCNetworkServiceRef
+}
+
+impl_TCFType!(
+    SCNetworkService,
+    SCNetworkServiceRef,
+    SCNetworkServiceGetTypeID
+);
+
+
+impl SCNetworkService {
+    /// Returns primary network service
+    pub fn global(store: &SCDynamicStore) -> Option<Self> {
+        if let Some(service_id) = global_query(store, "PrimaryService") {
+            for service in SCNetworkService::list() {
+                if service.id() == service_id {
+                    return Some(service);
+                }
+            }
+        }
+
+        return None;
+    }
+
+    /// Returns all available network services for the specified preferences.
+    pub fn list() -> Vec<SCNetworkService> {
+        let prefs = unsafe {
+            SCPreferencesCreate(
+                kCFAllocatorDefault,
+                CFString::from_static_string("ns_list").as_concrete_TypeRef(),
+                ptr::null(),
+            )
+        };
+
+        let array: CFArray<CFType> =
+            unsafe { CFArray::wrap_under_get_rule(SCNetworkServiceCopyAll(prefs)) };
+
+        array
+            .iter()
+            .map(|item| item.downcast::<SCNetworkService>().unwrap())
+            .collect::<Vec<SCNetworkService>>()
+    }
+
+    /// Returns the user-specified ordering of network services within the specified set.
+    pub fn list_order() -> Vec<SCNetworkService> {
+        let prefs = unsafe {
+            SCPreferencesCreate(
+                kCFAllocatorDefault,
+                CFString::from_static_string("ns_list_order").as_concrete_TypeRef(),
+                ptr::null(),
+            )
+        };
+
+        let netset = unsafe { SCNetworkSetCopyCurrent(prefs) };
+
+        let array: CFArray<CFType> =
+            unsafe { CFArray::wrap_under_get_rule(SCNetworkSetGetServiceOrder(netset)) };
+
+        let mut services = Vec::new();
+
+        for item in array.iter() {
+            if let Some(id) = item.downcast::<CFString>() {
+                let service_ref = unsafe {
+                    CFType::wrap_under_get_rule(SCNetworkServiceCopy(
```

Comment:
`SCNetworkServiceCopy` returns a `SCNetworkServiceRef` directly. So you can just do `SCNetworkService::wrap_under_get_rule(..)`, no need to go via `CFType` and downcast.

On the other hand, the function can return `NULL`. So you'll have to check for that before calling `wrap_under_get_rule`. The current code can reach undefined behavior if called on a non-existent service.